lib/api/bucketPutVersioning.js

```diff
@@ -0,0 +1,71 @@
+import { waterfall } from 'async';
+import { parseString } from 'xml2js';
+
+import metadata from '../metadata/wrapper';
+import services from '../services';
+
+/**
+ * Format of xml request:
+
+ <VersioningConfiguration xmlns="http://s3.amazonaws.com/doc/2006-03-01/">
+    <Status>VersioningState</Status>
+    <MfaDelete>MfaDeleteState</MfaDelete>
+ </VersioningConfiguration>
+
+ Note that there is the header in the request if setting MfaDelete:
+    x-amz-mfa: [SerialNumber] [TokenCode]
+ */
+
+/**
+ * Bucket Put Versioning - Create or update bucket Versioning
+ * @param {AuthInfo} authInfo - Instance of AuthInfo class with requester's info
+ * @param {object} request - http request object
+ * @param {object} log - Werelogs logger
+ * @param {function} callback - callback to server
+ * @return {undefined}
+ */
+export default function bucketPutVersioning(authInfo, request, log, callback) {
+    log.debug('processing request', { method: 'bucketPutVersioning' });
+
+    const bucketName = request.bucketName;
+    const metadataValParams = {
+        authInfo,
+        bucketName,
+        requestType: 'bucketPutVersioning',
+        log,
+    };
+
+    return waterfall([
+        next => services.metadataValidateAuthorization(metadataValParams,
+            (err, bucket) => next(err, bucket)), // to ignore null object
+        (bucket, next) => parseString(request.post, (err, result) => {
```

Comment:
I think it's not feasible to make them static the way you want. Look at the full functions of the second waterfall:
```js
(bucket, cb) => parseString(req, (err, res) => {
    cb();
});
```
- The first anonymous function uses an extra variable `req` apart from the inputs so we cannot make it a static function unless we use `req` as a global variable.
- The second anonymous function uses the extra variable `next` as well so it's difficult to make it static.
You have any other solution?

And also, practically the function `bucketGetVersioning` should not be a frequent operation.